lib/propolis/src/block/crucible.rs

```diff
@@ -109,34 +163,31 @@ impl block::Backend for CrucibleBackend {
         }
     }
 
-    fn attach(
-        &self,
-        dev: Arc<dyn block::Device>,
-        disp: &Dispatcher,
-    ) -> Result<()> {
-        let mut driverg = self.driver.lock().unwrap();
-        assert!(driverg.is_none());
-
-        let block_io = Arc::clone(&self.block_io);
-        let read_only = self.read_only;
-        let req_handler =
-            Box::new(move |req: &block::Request, ctx: &DispCtx| {
-                process_request(&*block_io, req, ctx, read_only)
-            });
-
-        // Spawn driver to service block dev requests
-        let driver = Arc::new(block::Driver::new(dev, req_handler));
-        driver.spawn("crucible", NonZeroUsize::new(8).unwrap(), disp)?;
-        *driverg = Some(driver);
+    fn process(&self, req: &block::Request, mem: &MemCtx) -> block::Result {
+        match self.process_request(req, mem) {
+            Ok(_) => block::Result::Success,
+            Err(e) => {
+                slog::info!(self.log, "block IO error {:?}", req.op; "error" => e);
+                block::Result::Failure
+            }
+        }
+    }
 
-        Ok(())
+    fn attach(&self, dev: Arc<dyn block::Device>) -> Result<()> {
+        self.driver.attach(dev)
     }
 }
 
 impl Entity for CrucibleBackend {
     fn type_name(&self) -> &'static str {
         "block-crucible"
     }
+    // TODO(#155): Move Crucible activation here. Also, this entity probably
+    // needs its own pause/resume routines akin to those in the other block
```

Comment:
Hmm.  There are some details here we have to discuss.  Specifically the block_size and volume_size are not known until after the volume is activated.  We may have to update the volume code to arrange the order here.

Reply:
https://github.com/oxidecomputer/crucible/issues/440

Reply:
I agree, this'll need some further untangling. Currently, the storage devices (NVMe/virtio-block) expect to get this information when they're created, so if we want to defer obtaining this information until after activation, we'll also need to change the device init contract accordingly.

I think this is possible in principle (skimming the code suggests that the devices only care once they're actually handling guest requests, and those will only arrive once `start` has been called, since the VM controller only starts vCPUs after starting entities), but I'm not sure how hairy the refactored code would be. We'd probably need to try it and see.

Reply:
I think taking our best guess on the Crucible side (as discussed in [issue 440](https://github.com/oxidecomputer/crucible/issues/440) ) is probably the easiest path forward here.